### PR TITLE
[GLUTEN-1500][VL] Implement OOM cap shared by tasks, and spill threshold shared by tasks and operators

### PR DESCRIPTION
**Pull request by zhztheplayer**
OOM cap shared by tasks: For cases like TPC-DS Q67
Spill threshold shared by tasks and operators: For cases like TPC-DS Q95

Previously, OOM cap was set for a single task, and spill threshold was set for a single operator.

Dependency https://github.com/oap-project/velox/pull/227

### CODE REVIEW

**Comment by zhztheplayer** [comment]
@zhejiangxiaomai 

The velox patch https://github.com/oap-project/velox/pull/227 might be a little bit complicated to rebase during your next Velox upgrading action.

I can give some help then if needed.

**Comment by zhztheplayer** [comment]
Will fix a avg decimal spill bug triggered by this PR in https://github.com/oap-project/velox/pull/236

File: cpp/velox/compute/VeloxBackend.cc [hunk 1]
```diff
@@ -56,7 +57,7 @@ VeloxBackend::VeloxBackend(const std::unordered_map<std::string, std::string>& c
   // mem tracker
   int64_t maxMemory;
   {
-    auto got = confMap_.find(kSparkTaskOffHeapMemory); // per task, for creating iterator
+    auto got = confMap_.find(kSparkOffHeapMemory); // per executor, shared by tasks for creating iterator
```
Comment (winningsix) [inline cpp/velox/compute/VeloxBackend.cc:60]: Question: [```memPoolOptions_```](https://github.com/oap-project/gluten/blob/5c7d94a3866c8f927ba114af8400f0f04022ff45/cpp/velox/memory/VeloxMemoryPool.cc#L426) will be used when calling [```GetResultIterator```](https://github.com/oap-project/gluten/blob/5c7d94a3866c8f927ba114af8400f0f04022ff45/cpp/velox/compute/VeloxBackend.cc#LL95C47-L95C64) by each task. How is it shared by tasks?

File: docs/get-started/Velox.md [hunk 2]
```diff
@@ -366,11 +366,11 @@ The following tips may help when you are trying to tune against spilling-related
...
+  This is to set the memory cap limit (represented as MEMORY_LIMIT) for Velox backend. The formula is: MEMORY_LIMIT = offHeap.size * memoryCapRatio. OOM will be raised if the memory usage exceeds this limit.
 
 * `spark.gluten.sql.columnar.backend.velox.spillMemoryThresholdRatio`
 
-  This is to set the spill limit (represented as SPILL_LIMIT) to each operator in each Velox task. The formula is: SPILL_LIMIT = MEMORY_LIMIT * spillMemoryThresholdRatio. Once a single Velox operator has allocated memory larger than this size, a spill request will be sent to the operator to force spilling-to-disk procedure before it continues processing. 
+  This is to set the spill limit (represented as SPILL_LIMIT) for Velox operators during execution. The formula is: SPILL_LIMIT = MEMORY_LIMIT * spillMemoryThresholdRatio. Once the allocated memory by Velox becomes larger than this size, a spill request will be sent to the current active operator to force spilling-to-disk procedure before it continues processing.
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply docs/get-started/Velox.md:373]: > will "active Velox tasks" or "Velox memory management system" be more suitable?

It sounds great but as of now Velox doesn't implement coordinated spilling across the whole task, for its current version. But they are working on it:
https://github.com/facebookincubator/velox/pulls/xiaoxmeng

So I guess we can change to the form "active Velox tasks" or "Velox memory management system" once we upgrade to newer version of Velox?

File: cpp/velox/compute/WholeStageResultIterator.cc [hunk 3]
```diff
@@ -220,21 +217,17 @@ void WholeStageResultIterator::setConfToQueryContext(const std::shared_ptr<velox
...
-    auto defaultSpillThreshold = std::to_string((long)(defaultSpillThresholdRatio * (double)maxMemory));
-
     // Spill configs
     configs[velox::core::QueryConfig::kSpillEnabled] = getConfigValue(kSpillEnabled, "true");
     configs[velox::core::QueryConfig::kAggregationSpillEnabled] = getConfigValue(kAggregationSpillEnabled, "true");
     configs[velox::core::QueryConfig::kJoinSpillEnabled] = getConfigValue(kJoinSpillEnabled, "true");
     configs[velox::core::QueryConfig::kOrderBySpillEnabled] = getConfigValue(kOrderBySpillEnabled, "true");
     configs[velox::core::QueryConfig::kAggregationSpillMemoryThreshold] =
-        getConfigValue(kAggregationSpillMemoryThreshold, defaultSpillThreshold); // spill only when input doesn't fit
+        getConfigValue(kAggregationSpillMemoryThreshold, "0"); // spill only when input doesn't fit
```
[the remark this answers is not in the document]
Comment (zhztheplayer) [reply cpp/velox/compute/WholeStageResultIterator.cc:226]: "0" here means "spill when input doesn't fit". 

The PR is to help Velox make the decision of "input doesn't fit" automatically with some predefined cap numbers passed to Velox.

After this PR (or probably with some follow-up optimizations) I expect that the overall spilling system can become optimal with just default settings. But in some rare cases user may still need these threshold settings to tune against the system.

File: cpp/velox/compute/VeloxBackend.cc (hunk 1 above)
Comment (zhztheplayer) [reply cpp/velox/compute/VeloxBackend.cc:60]: Previously we passed the option to each task but in the PR I have changed to this way:
https://github.com/zhztheplayer/gluten/blob/4e8b486cf72546e5b414c560bf227a286990fcf7/cpp/velox/memory/VeloxMemoryPool.cc#L455-L465

The process-wise static pool will be the parent pool of all context pools for the tasks, so the cap will be shared by the tasks.

**Review by zhouyuan — APPROVED** [review]
👍

**Pull request merged**